File: lab5/c/Canvas.c
//
//  Canvas.c
//
//  Created by Warren R. Carithers 2016/11/03.
//  Based on a C++ version created by Joe Geigel.
//  Copyright 2016 Rochester Institute of Technology.  All rights reserved.
//
//  Routines for adding points to create a new mesh.
//
//  This file should not be modified by students.
//

#if defined(_WIN32) || defined(_WIN64)
#include <windows.h>
#endif

#ifdef __APPLE__
#include <OpenGL/gl.h>
#else
#include <GL/glew.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include "Canvas.h"
#include "floatVector.h"

///
// Constructor
//
// @param w width of canvas
// @param h height of canvas
///
Canvas *Canvas_create( int w, int h ) {
    Canvas *C = calloc( 1, sizeof(Canvas) );
    if( C != NULL ) {
        C->width = w;
        C->height = h;
        // all other fields contain zeroes
    }
    return( C );
}


///
// Destructor
//
// @param C canvas to destroy
///
void Canvas_destroy( Canvas *C ) {
    if( !C )
        return;
    free( C );
}


///
// clear the canvas
//
// @param C which canvas to clear
///
void Canvas_clear( Canvas *C )
{
    if( C->pointArray ) {
        free( C->pointArray );
        C->pointArray = 0;
    }
    if( C->elemArray ) {
        free( C->elemArray );
        C->elemArray = 0;
    }
    floatVectorClear( &(C->points) );
    C->numElements = 0;
}


///
// adds a triangle to the current shape
//
// @param C The Canvas to be used
// @param p0 first triangle vertex
// @param p1 second triangle vertex
// @param p2 final triangle vertex
///
void Canvas_addTriangle( Canvas *C, Vertex p0, Vertex p1, Vertex p2 )
{
    floatVectorPushBack( &(C->points), p0.x );
    floatVectorPushBack( &(C->points), p0.y );
    floatVectorPushBack( &(C->points), p0.z );
    floatVectorPushBack( &(C->points), 1.0f );

    floatVectorPushBack( &(C->points), p1.x );
    floatVectorPushBack( &(C->points), p1.y );
    floatVectorPushBack( &(C->points), p1.z );
    floatVectorPushBack( &(C->points), 1.0f );

    floatVectorPushBack( &(C->points), p2.x );
    floatVectorPushBack( &(C->points), p2.y );
    floatVectorPushBack( &(C->points), p2.z );
    floatVectorPushBack( &(C->points), 1.0f );

    C->numElements += 3;  // three vertices per triangle
}


///
// gets the array of vertices for the current shape
//
// @param C The Canvas to use
///
float *Canvas_getVertices( Canvas *C )
{
    // delete the old point array if we have one
    if( C->pointArray ) {
        free( C->pointArray );
    }

    // create and fill a new point array
    C->pointArray = (float *) malloc(
        floatVectorSize(&(C->points)) * sizeof(float) );
    if( C->pointArray == 0 ) {
    	perror( "point allocation failure" );
	exit( 1 );
    }
    for( int i = 0; i < floatVectorSize(&(C->points)); i++ ) {
        C->pointArray[i] = C->points.vec[i];
    }

    return C->pointArray;
}


///
// gets the array of elements for the current shape
//
// @param C The Canvas to use
///
GLuint *Canvas_getElements( Canvas *C )
{
    // delete the old point array if we have one
    if( C->elemArray ) {
        free( C->elemArray );
    }

    // create and fill a new element array
    C->elemArray = (GLuint *) malloc( C->numElements * sizeof(GLuint) );
    if( C->elemArray == 0 ) {
    	perror( "element allocation failure" );
	exit( 1 );
    }
    for( int i = 0; i < C->numElements; i++ ) {
        C->elemArray[i] = i;
    }

    return C->elemArray;
}


///
// returns number of vertices in current shape
//
// @param C The Canvas to use
///
int Canvas_nVertices( Canvas *C )
{
    return C->numElements;
}

File: lab4/java/Canvas.java
//
//  Canvas.java
//
//  Created by Warren R. Carithers on 2016/11/03.
//  Copyright 2016 Rochester Institute of Technology.  All rights reserved.
//
//  This file should not be modified by students.
//

import java.awt.*;
import java.nio.*;
import java.awt.event.*;
import javax.media.opengl.*;
import javax.media.opengl.awt.GLCanvas;
import java.util.*;

public class Canvas {

    ///
    // canvas size information
    ///
    private int width;
    private int height;

    ///
    // point-related data
    ///
    private Vector<Float> points;
    private int numElements;
    private Vector<Integer> elements;

    ///
    // Constructor
    //
    // @param w width of canvas
    // @param h height of canvas
    ///
    Canvas( int w, int h )
    {
        width = w;
        height = h;
	clear();    // do this the easy way
    }

    ///
    // clear the canvas
    ///
    public void clear()
    {
        // easiest method:  just allocate new ones!
	numElements = 0;
        points = new Vector<Float>();
        elements = new Vector<Integer>();
    }


    ///
    // adds a triangle to the current shape
    //
    // @param p0 first triangle vertex
    // @param p1 second triangle vertex
    // @param p2 final triangle vertex
    ///
    public void addTriangle( Vertex p0, Vertex p1, Vertex p2 )
    {
        points.add( new Float(p0.x) );
        points.add( new Float(p0.y) );
        points.add( new Float(p0.z) );
        points.add( new Float(1.0f) );

        points.add( new Float(p1.x) );
        points.add( new Float(p1.y) );
        points.add( new Float(p1.z) );
        points.add( new Float(1.0f) );

        points.add( new Float(p2.x) );
        points.add( new Float(p2.y) );
        points.add( new Float(p2.z) );
        points.add( new Float(1.0f) );

        numElements += 3;  // three vertices per triangle
    }

    ///
    // get the array of vertices for the current shape
    ///
    public Buffer getVertices()
    {
        float v[] = new float[points.size()];
        for( int i=0; i < points.size(); i++ ) {
            v[i] = (points.elementAt(i)).floatValue();
        }
        return FloatBuffer.wrap( v );
    }

    ///
    // get the array of elements for the current shape
    ///
    public Buffer getElements()
    {
        int e[] = new int[elements.size()];
        for( int i=0; i < elements.size(); i++ ) {
            e[i] = (elements.elementAt(i)).intValue();
        }
        return IntBuffer.wrap( e );
    }

    ///
    // get the number of vertices in the current shape
    ///
    public int nVertices()
    {
        return numElements;
    }

}
